# Advanced Dataframes Exercises I

### 1. Run python -m pip install mysqlclient pymysql from your terminal to install the mysql client (any folder is fine)

### 2. cd into your exercises folder for this module and run echo env.py >> .gitignore

### 3. Create a function named `get_db_url`. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

In [2]:
from env import user, password, host
import pandas as pd
import numpy as np

def get_db_url(username: str, hostname: str , password: str, database_name: str):
    '''
    Takes username, hostname, password and database_name and 
    returns a connection string
    '''
    connection = f'mysql+pymysql://{username}:{password}@{hostname}/{database_name}'
    
    return connection

emp_conn = get_db_url(user, host, password, 'employees')


### 4. Use your function to obtain a connection to the employees database.

In [181]:
sql = '''
select *
from employees
limit 5
offset 50
'''
pd.read_sql(sql, emp_conn)

emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

### 5a. Intentionally make a typo in the database url. What kind of error message do you see? 

* Returns: ```NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:mysqltest.pymysql
```

In [254]:
bad_conn = emp_conn[:5] + 'test' + emp_conn[5:]
# pd.read_sql(sql, bad_conn)

### 5b. Intentionally make an error in your SQL query. What does the error message look like? 

* Returns: ```ProgrammingError: (pymysql.err.ProgrammingError) (1146, "Table 'employees.employeeeeeeeeees' doesn't exist")```

In [253]:
bad_sql = '''
select *
from employeeeeeeeeees
limit 5
offset 50
'''
# pd.read_sql(bad_sql, emp_conn)

### 6. Read the employees and titles tables into two separate DataFrames.

In [4]:
sql_employees = '''
select *
from employees
'''
employees = pd.read_sql(sql_employees, emp_conn)
sql_titles = '''
select *
from titles
'''
titles = pd.read_sql(sql_titles, emp_conn)

### 7. How many rows and columns do you have in each DataFrame? Is that what you expected? 

* 300024 rows, 6 columns - employees / 443308 rows, 4 columns and Yes because there would be more titles than employees as each employee will have at least one title and can have many. Yes this is what I expected.

In [255]:
[employees.shape, titles.shape]

[(300024, 6), (443308, 4)]

### 8. Display the summary statistics for each DataFrame.

In [29]:
employees.describe()
titles.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

### 9. How many unique titles are in the `titles` DataFrame?

In [256]:
len(titles.title.unique())

7

### 10. What is the oldest date in the to_date column?

In [37]:
titles.to_date.min()

datetime.date(1985, 3, 1)

### 11. What is the most recent date in the to_date column?

In [59]:
titles[titles.to_date != datetime.date(9999, 1, 1)].to_date.max()

datetime.date(2002, 8, 1)

# Exercises II

### 1. Copy the `users` and `roles` DataFrames from the examples above.

In [61]:
users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})


### 2. What is the result of using a `right` join on the DataFrames?



In [80]:
users.join(roles, how='right', lsuffix='_users', rsuffix='_roles')


id_users name_users  role_id  id_roles name_roles
0         1        bob      1.0         1      admin
1         2        joe      2.0         2     author
2         3      sally      3.0         3   reviewer
3         4       adam      3.0         4  commenter

### 3. What is the result of using an outer join on the DataFrames?

In [82]:
users.join(roles, how='outer', lsuffix='_users', rsuffix='_roles')

id_users name_users  role_id  id_roles name_roles
0         1        bob      1.0       1.0      admin
1         2        joe      2.0       2.0     author
2         3      sally      3.0       3.0   reviewer
3         4       adam      3.0       4.0  commenter
4         5       jane      NaN       NaN        NaN
5         6       mike      NaN       NaN        NaN

### 4. What happens if you drop the foreign keys from the DataFrames and try to merge them?

In [90]:
users.drop(columns='role_id')
users.merge(roles, how='outer')

id       name  role_id
0   1        bob      1.0
1   2        joe      2.0
2   3      sally      3.0
3   4       adam      3.0
4   5       jane      NaN
5   6       mike      NaN
6   1      admin      NaN
7   2     author      NaN
8   3   reviewer      NaN
9   4  commenter      NaN

### 5. Load the `mpg` dataset from PyDataset.

In [94]:
from pydataset import data
mpg = data('mpg')

### 6. Output and read the documentation for the `mpg` dataset.



In [95]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




### 7. How many rows and columns are in the dataset?

* 234 Rows and 11 Columns

In [96]:
mpg.shape

(234, 11)

### 8. Check out your column names and perform any cleanup you may want on them.

In [100]:
mpg.drop(columns=['displ', 'fl', 'cyl', 'drv'])

manufacturer   model  year       trans  cty  hwy    class
1           audi      a4  1999    auto(l5)   18   29  compact
2           audi      a4  1999  manual(m5)   21   29  compact
3           audi      a4  2008  manual(m6)   20   31  compact
4           audi      a4  2008    auto(av)   21   30  compact
5           audi      a4  1999    auto(l5)   16   26  compact
..           ...     ...   ...         ...  ...  ...      ...
230   volkswagen  passat  2008    auto(s6)   19   28  midsize
231   volkswagen  passat  2008  manual(m6)   21   29  midsize
232   volkswagen  passat  1999    auto(l5)   16   26  midsize
233   volkswagen  passat  1999  manual(m5)   18   26  midsize
234   volkswagen  passat  2008    auto(s6)   17   26  midsize

[234 rows x 7 columns]

### 9. Display the summary statistics for the dataset.

In [101]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

### 10. How many different manufacturers are there?

* There are 15 different manufacturers.

In [105]:
len(mpg.manufacturer.unique())

15

### 11. How many different models are there?

* There are 38 different models

In [107]:
len(mpg.model.unique())

38

### 12. Create a column named `mileage_difference` like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

In [111]:
mpg['mileage_difference'] = mpg.hwy - mpg.cty

### 13. Create a column named `average_mileage` like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

In [114]:
mpg['average_mileage'] = round(mpg[['cty', 'hwy']].mean(axis=1), 2)

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

### 14. Create a new column on the `mpg` dataset named `is_automatic` that holds boolean values denoting whether the car has an automatic transmission.

In [127]:
mpg['is_automatic'] = mpg.trans.str.contains('auto')

### 15. Using the `mpg` dataset, find out which which manufacturer has the best miles per gallon on average?

|manufacturer|model|displ|year|cyl|trans|drv|cty|hwy|fl|class|mileage_difference|average_mileage|is_automatic|
|------|------|------|------|------|------|------|------|------|------|------|------|------|------|
|222|volkswagen|new beetle|1.9|1999|4|manual(m5)|f|35|44|d|subcompact|9|39.5|False|


In [130]:
mpg[mpg.average_mileage == mpg.average_mileage.max()]

manufacturer       model  displ  year  cyl       trans drv  cty  hwy fl  \
222   volkswagen  new beetle    1.9  1999    4  manual(m5)   f   35   44  d   

          class  mileage_difference  average_mileage  is_automatic  
222  subcompact                   9             39.5         False

### 16. Do automatic or manual cars have better miles per gallon? -- 

* Manual's have better mpg in this data set on average:
|auto trans|manual trans|
|---|---|
|19.130573248407643|22.227272727272727|

In [136]:
auto_mpg_avg = mpg[mpg.trans.str.contains('auto')].average_mileage.mean()
manual_mpg_avg = mpg[mpg.trans.str.contains('manual')].average_mileage.mean()
auto_mpg_avg, manual_mpg_avg

(19.130573248407643, 22.227272727272727)

# Exercises III

### 1. Use your `get_db_url` function to help you explore the data from the chipotle database.

In [24]:
conn = get_db_url(username=user, password=password, hostname=host, database_name='chipotle')
sql = '''
select *
from orders'''
chipotle = pd.read_sql(sql, conn)
chipotle

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   nan     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   nan     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 6 columns]

### 2. What is the total price for each order?

* $34,500.16 is the total price for all orders.

In [25]:
chipotle.item_price.replace('[\$,]', '', regex=True).astype(float).sum()

34500.16

### 3. What are the most popular 3 items?

* The three most popular items are:

<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>id</th>      <th>order_id</th>      <th>quantity</th>      <th>item_name</th>      <th>choice_description</th>      <th>item_price</th>    </tr>  </thead>  <tbody>    <tr>      <th>3598</th>      <td>3599</td>      <td>1443</td>      <td>15</td>      <td>Chips and Fresh Tomato Salsa</td>      <td>nan</td>      <td>44.25</td>    </tr>    <tr>      <th>3887</th>      <td>3888</td>      <td>1559</td>      <td>8</td>      <td>Side of Chips</td>      <td>nan</td>      <td>13.52</td>    </tr>    <tr>      <th>4152</th>      <td>4153</td>      <td>1660</td>      <td>10</td>      <td>Bottled Water</td>      <td>nan</td>      <td>15.00</td>    </tr>  </tbody></table>


In [49]:
bools = chipotle.quantity.isin(chipotle.quantity.sort_values(ascending=False).head(n=3))
chipotle[bools]

id  order_id  quantity                     item_name  \
3598  3599      1443        15  Chips and Fresh Tomato Salsa   
3887  3888      1559         8                 Side of Chips   
4152  4153      1660        10                 Bottled Water   

     choice_description item_price  
3598                nan    $44.25   
3887                nan    $13.52   
4152                nan    $15.00

### 4. Which item has produced the most revenue?

<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>id</th>      <th>order_id</th>      <th>quantity</th>      <th>item_name</th>      <th>choice_description</th>      <th>item_price</th>      <th>revenue</th>    </tr>  </thead>  <tbody>    <tr>      <th>3598</th>      <td>3599</td>      <td>1443</td>      <td>15</td>      <td>Chips and Fresh Tomato Salsa</td>      <td>nan</td>      <td>44.25</td>      <td>663.75</td>    </tr>  </tbody></table>

In [41]:
mxidx = chipotle.item_price.replace('[\$,]', '', regex=True).astype(float).idxmax()
chipotle.loc[mxidx]

id                                            3599
order_id                                      1443
quantity                                        15
item_name             Chips and Fresh Tomato Salsa
choice_description                             nan
item_price                                 $44.25 
Name: 3598, dtype: object

### 5. Join the `employees` and `titles` DataFrames together.

In [45]:
emp_w_title = employees.merge(titles, how='left')
emp_w_title

emp_no  birth_date first_name last_name gender   hire_date  \
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
...        ...         ...        ...       ...    ...         ...   
443303  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443304  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443305  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443306  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443307  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30   

                  title   from_date     to_date  
0       Senior Engineer  1986-06-26  9999-01-01  
1                 Staff  1996-08-03  9999-01-01  
2       Senior Engineer  1995-12-03  9999-01-01  
3              Engineer  1986-12-01  1995-12-01  
4       Senior Engineer  1995-12-01  9999-01-01  
...                 ...         ...         ...  
443303         Engineer  1987-08-30  1992-08-29  
443304  Senior Engineer  1992-08-29  9999-01-01  
443305     Senior Staff  1998-12-27  9999-01-01  
443306            Staff  1993-12-27  1998-12-27  
443307         Engineer  1997-11-30  9999-01-01  

[443308 rows x 9 columns]

### 6. For each title, find the hire date of the employee that was hired most recently with that title.

In [47]:
hire_dte = emp_w_title.groupby(by='title').hire_date.agg(['max'])
hire_dte.rename(columns={'max': 'most recent hire date'})

most recent hire date
title                                   
Assistant Engineer            1999-12-24
Engineer                      2000-01-28
Manager                       1992-02-05
Senior Engineer               2000-01-01
Senior Staff                  2000-01-13
Staff                         2000-01-12
Technique Leader              1999-12-31

### 7. Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

In [48]:
dept_sql = '''
select title, dept_name
from employees
join dept_emp
using(emp_no)
join departments
using(dept_no)
join titles
on titles.emp_no = employees.emp_no
and titles.to_date >= now() #You could remove this and get all titles
'''
title_dept = pd.read_sql(dept_sql, emp_conn)
# title_dept.groupby(['dept_name','title']).title.agg(['count'])
pd.crosstab(title_dept.dept_name, title_dept.title, margins=True).T


dept_name           Customer Service  Development  Finance  Human Resources  \
title                                                                         
Assistant Engineer                68         1833        0                0   
Engineer                         627        15677        0                0   
Manager                            1            1        1                1   
Senior Engineer                 1790        43364        0                0   
Senior Staff                   12349         1085    10650            10843   
Staff                           3902          315     3199             3416   
Technique Leader                 241         6117        0                0   
All                            18978        68392    13850            14260   

dept_name           Marketing  Production  Quality Management  Research  \
title                                                                     
Assistant Engineer          0        1542                 433        77   
Engineer                    0       13325                3744       830   
Manager                     1           1                   1         1   
Senior Engineer             0       37156               10390      2250   
Senior Staff            12371        1123                   0     10219   
Staff                    3880         349                   0      3206   
Technique Leader            0        5210                1422       321   
All                     16252       58706               15990     16904   

dept_name           Sales     All  
title                              
Assistant Engineer      0    3953  
Engineer                0   34203  
Manager                 1       9  
Senior Engineer         0   94950  
Senior Staff        32032   90672  
Staff                9967   28234  
Technique Leader        0   13311  
All                 42000  265332